# Advent of Code 2020 - day 3

## Challenge

---------

In [2]:
from aocd import data


In [6]:
data = list(data.split('\n'))

In [7]:
data

['.#......#..####.....#..#.......',
 '#.#...#...#..#.#...#.#...##.##.',
 '#.#....#..........#...##.....##',
 '#.#.#.....##......#.#.......###',
 '..#..###....#.#....#.#.#..#....',
 '.......#.#....##..##...#...#...',
 '..#..#..#..###.......#.....#.#.',
 '.#.......#...##...##.##......##',
 '#.#.##..##.#..#....#..###..#.#.',
 '#.....#.#.........#.....##.#.#.',
 '..#.#....##..#...#...##........',
 '......#....#..##.#.#......###..',
 '.......#.......#......##...#...',
 '.##.....#.......#...###.....##.',
 '.#...#.##..##.#..##....#.......',
 '..#......##...#..#...#.#.##.###',
 '.##.##.....##....#..#......#.#.',
 '.#.....#..###..#.##.#.....##.#.',
 '......##..........#..........#.',
 '.##....#.....#..##.#..#.#..###.',
 '..##.......#....#...##...#..#..',
 '.##...#.....#.###.#.#..#...#.#.',
 '.....##.#.##..##...#...........',
 '..#..###.##.#.#.###...###..#.#.',
 '.#........#..#.#........#.#...#',
 '....##.......#....#.#.##.#.....',
 '....##........######..###..#.#.',
 '#.#.#............#.......#

--- Day 3: Toboggan Trajectory ---
With the toboggan login problems resolved, you set off toward the airport. While travel by toboggan might be easy, it's certainly not safe: there's very minimal steering and the area is covered in trees. You'll need to see which angles will take you near the fewest trees.

Due to the local geology, trees in this area only grow on exact integer coordinates in a grid. You make a map (your puzzle input) of the open squares (.) and trees (#) you can see. For example:

`..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#`

These aren't the only trees, though; due to something you read about once involving arboreal genetics and biome stability, the same pattern repeats to the right many times:

`..##.........##.........##.........##.........##.........##.......  --->
#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..
.#....#..#..#....#..#..#....#..#..#....#..#..#....#..#..#....#..#.
..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#
.#...##..#..#...##..#..#...##..#..#...##..#..#...##..#..#...##..#.
..#.##.......#.##.......#.##.......#.##.......#.##.......#.##.....  --->
.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#
.#........#.#........#.#........#.#........#.#........#.#........#
#.##...#...#.##...#...#.##...#...#.##...#...#.##...#...#.##...#...
#...##....##...##....##...##....##...##....##...##....##...##....#
.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#.#..#...#.#  --->`

You start on the open square (.) in the top-left corner and need to reach the bottom (below the bottom-most row on your map).

The toboggan can only follow a few specific slopes (you opted for a cheaper model that prefers rational numbers); start by counting all the trees you would encounter for the slope right 3, down 1:

From your starting position at the top-left, check the position that is right 3 and down 1. Then, check the position that is right 3 and down 1 from there, and so on until you go past the bottom of the map.

The locations you'd check in the above example are marked here with O where there was an open square and X where there was a tree:

`..##.........##.........##.........##.........##.........##.......  --->
#..O#...#..#...#...#..#...#...#..#...#...#..#...#...#..#...#...#..
.#....X..#..#....#..#..#....#..#..#....#..#..#....#..#..#....#..#.
..#.#...#O#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#..#.#...#.#
.#...##..#..X...##..#..#...##..#..#...##..#..#...##..#..#...##..#.
..#.##.......#.X#.......#.##.......#.##.......#.##.......#.##.....  --->
.#.#.#....#.#.#.#.O..#.#.#.#....#.#.#.#....#.#.#.#....#.#.#.#....#
.#........#.#........X.#........#.#........#.#........#.#........#
#.##...#...#.##...#...#.X#...#...#.##...#...#.##...#...#.##...#...
#...##....##...##....##...#X....##...##....##...##....##...##....#
.#..#...#.#.#..#...#.#.#..#...X.#.#..#...#.#.#..#...#.#.#..#...#.#  --->`

In this example, traversing the map using this slope would cause you to encounter 7 trees.

Starting at the top-left corner of your map and following a slope of right 3 and down 1, how many trees would you encounter?

In [6]:
# Prepare an empty list for our data
data = []

# Open the file as f, pull one line at a time, strip it of the \n at the end of each line, and append it to our data. 
with open("AoC2020_3.txt") as f:
    for line in f.readlines():
        data.append(line.strip())

In [9]:
print(len(data))
print(len(data[0]))

323
31


Our map is 323 entries long and 31 entries wide. 

For the first challenge of a right 3, down 1 slope, we will need 323 steps to reach the bottom. 

In [43]:
j = 0
trees = 0
for i in range(0, len(data)):
    
    j = j % len(data[0])
    
    if data[i][j] == '#':
        trees += 1

    j = j + 7

In [44]:
trees

99

3/1 = 151 trees - correct
1/1 = 103
5/1 = 83
7/1 = 99
1/2 = 59

In [42]:
151*103*83*99*59

7540141059